pvl/pvl_validate.py

```diff
@@ -0,0 +1,218 @@
+# -*- coding: utf-8 -*-
+"""A program for testing and validating PVL text.
+
+The ``pvl_validate`` program will read a file with PVL text (any of
+the kinds of files that :func:`pvl.load` reads) and will report
+on which of the various PVL dialects were able to load that PVL
+text, and then also reports on whether the ``pvl`` library can encode
+the Python Objects back out to PVL text.
+
+You can imagine some PVL text that could be loaded, but is not able
+to be written out in a particular strict PVL dialect (like PDS3
+labels).
+"""
+
+# Copyright 2020, ``pvl`` library authors.
+#
+# Reuse is permitted under the terms of the license.
+# The AUTHORS file and the LICENSE file are at the
+# top level of this library.
+
+import argparse
+import logging
+import sys
+
+import pvl
+from .lexer import LexerError
+from .grammar import PVLGrammar, ODLGrammar, ISISGrammar, OmniGrammar
+from .parser import ParseError, PVLParser, ODLParser, OmniParser
+from .decoder import PVLDecoder, ODLDecoder, OmniDecoder
+from .encoder import PVLEncoder, ODLEncoder, ISISEncoder, PDSLabelEncoder
+
+dialects = dict(PDS3=dict(parser=ODLParser(),
```

Comment:
Although this is fine for now, if these dialect objects are considered immutable objects it could be more appropriate to use dataclasses or named tuples (https://docs.python.org/3/library/dataclasses.html), or at least the dialects at the top level can be a dict, but then each dialect would be a dialect dataclass, that way you can access the various sub parts using properties instead of key strings e.g. `Dialect.parser`. Alternatively a dialect class type could be defined that must have a parser class, a grammar class, etc could be used. I would say that this is not a blocker for merger but worth considering as a enhancement afterwards to make this area a bit more typed/OO

Reply:
So *dataclasses* are a new feature in Python 3.7, to keep this 3.6-compliant, we can't use them.  Seems like *dataclasses* are just syntactic sugar for writing a simple class, so obviously this can be done with just the existing Python class mechanism.

Actually leaving the deeper elements as dicts allows easy passing of them to the ``pvl.loads`` and ``pvl.dumps`` functions via the ``**`` operator in the ``pvl_flavor()`` function.  I don't think there's a big improvement by trying to go to classes here.  But sure, this can be handled as a subsequent enhancement.

Reply:
Just FYI, there's a semi-official backport of dataclasses to 3.6, from the author of the dataclasses https://github.com/ericvsmith/dataclasses. I'm not advocating necessarily for the use of dataclasses here, but I strongly disagree that they are only syntactic sugar, simply because if I need to write or read a variable name 3 times or once is a huge difference for scanning a class for bugs.

Reply:
Yeah, no need to use this backport for a single use here.

However, it also seems like you think 'syntactic sugar' is a negative term, that's not how I think of it. As you say, it is a way to more clearly express a complicated language construct, which is a good thing.

Reply:
sounds good